File: leetcode/permutattions/permutattions.cpp
#include <vector>

using std::vector;

class Solution {
private:
    int limit;

public:
    vector<vector<int>> permute(vector<int>& nums) {
        limit = nums.size();
        vector<vector<int>> ans;
        vector<int> currAns;
        fn(ans, currAns, nums, 0);
        return ans;
    }

    void fn(vector<vector<int>>& ans, vector<int>& currAns, vector<int>& nums, int targetIndex) {
        if(currAns.size() >= limit) {
            ans.push_back(vector<int>(currAns));
            return;
        }

        currAns.push_back(nums[targetIndex]);
        fn(ans, currAns, nums, targetIndex+1);
        for(int i = targetIndex; i > 0; i--) {
            int tmp = currAns[i];
            currAns[i] = currAns[i-1];
            currAns[i-1] = tmp;
            fn(ans, currAns, nums, targetIndex+1);
        }
        currAns.erase(currAns.begin());
    }
};

File: leetcode/permutattions/permutattions.java
package permutattions;

import java.util.ArrayList;
import java.util.List;

class Solution {
    private List<List<Integer>> ans;
    private int maxAllowed;
    private int[] nums;

    public List<List<Integer>> permute(int[] nums) {
        this.nums = nums;
        this.maxAllowed = nums.length;
        this.ans = new ArrayList<>();
        List<Integer> currPerm = new ArrayList<>();
        fn(currPerm, 0);
        return ans;
    }

    private void fn(List<Integer> currPerm, int targetIndex) {
        if(currPerm.size() >= maxAllowed) {
            ans.add(new ArrayList<>(currPerm));
            return;
        }

        currPerm.add(nums[targetIndex]);
        fn(currPerm, targetIndex+1);
        for(int i = targetIndex; i > 0; i--) {
            int tmp = currPerm.get(i);
            currPerm.set(i, currPerm.get(i-1));
            currPerm.set(i-1, tmp);
            fn(currPerm, targetIndex+1);
        }
        currPerm.remove(0);
    }
}
